### PR TITLE
V2 api

### PR DESCRIPTION
**Pull request by BenMorganIO**
- [x] Abstract
- [x] Motivation
- [x] Detailed Design
- [x] Drawbacks
- [x] Alternatives
- [x] Unresolved Questions

### CODE REVIEW

**Comment by BenMorganIO** [comment]
@yurifrl I'll be editing this draft a lot today. I've added everything in the prospective of helping Ember and making things fast-ish. The current draft does not take the JSON API into account.

**Comment by yurifrl** [comment]
Ok, just let me know, i'm in a process of creating a ember test app, so i will take any problem a have in consideration as well, i'm also tweaking v2

**Comment by BenMorganIO** [comment]
There's a lot of work left. You'll need to add the new [JSON API header](http://www.iana.org/assignments/media-types/application/vnd.api+json) and hopefully find a way to get a simple XML response somehow.

**Comment by yurifrl** [comment]
i've been looking throw xml in ams, and it's not good https://github.com/rails-api/active_model_serializers/search?q=xml&type=Issues&utf8=%E2%9C%93

**Comment by yurifrl** [comment]
also, are we willing to stop using nested resources? (oh lord, just image to re-factor all that code)

http://discuss.emberjs.com/t/i-was-being-silly-to-fight-ember-data-on-nested-urls-i-think/2082

**Comment by BenMorganIO** [comment]
The first release candidate of [Active Model Serializers 0.10](https://github.com/rails-api/active_model_serializers/releases/tag/v0.10.0.rc1) has been released.

**Comment by yurifrl** [comment]
shall we use it?

**Comment by BenMorganIO** [comment]
@yurifrl I've added that to the Unresolved Questions section.

**Comment by futhr** [comment]
+1 Why not also add notes about those asset related dependencies that do not really belong in spree core to make the use of a strict spree api service as light weight as possible.

**Comment by BenMorganIO** [comment]
@futhr I think that's a spree_core issue. That seems to be only a small issue. The [logo](https://github.com/spree/spree/blob/master/core/app/assets/images/logo/spree_50.png) and the [no-images](https://github.com/spree/spree/tree/master/core/app/assets/images/noimage) are both used by spree_backend and the spree_frontend. I don't think it actually has anything to with the spree_api, its more of an unfortunate side effect.

**Comment by BenMorganIO** [comment]
Its only ~30 or so KB for all the images. I wouldn't worry too much.

**Comment by hhff** [comment]
Hi Guys!

@JDutil pointed me this way via Twitter!  @BenMorganIO - been checking your posts about Spree's API.  You're doing awesome work man!

I just released http://www.spree-ember.com in the last day.  It's a set of 4 ember addons for building Ember Apps against Spree's API.

A big part of making it possible was by building this gem:
https://github.com/hhff/spree_ams

It's super simple - it adds an "ams" namespace to the `api` routes, ala `localhost:3000/api/ams/products.json`

All the controllers inherit from the regular API controllers.

Check out a sample response here: http://testing.spree-ember.com/api/ams/products

I've also done a lot of the work actually modeling the data in Ember Data: https://github.com/hhff/spree-ember/tree/master/packages/core/app/models

Going to have a big read of this RFC and post some thoughts - I'd love to be of whatever help I can!  Just wanted to say Hi and link to the above projects for your own reference / interest :)

hx

**Comment by JDutil** [comment]
@hhff while I don't expect everything to just be accepted outright I would imagine that using spree_ams under the v2 namespace rather than ams namespace would probably be a good start.

**Comment by BenMorganIO** [comment]
@hhff THANK YOU for the ember code :heart:

The great part about Ember is the adapters. What's so cool is that you can if the data in AMS switches out for the JSON Spec, there's an adapter for that!

The good news is 0.10.rc1 is released. This should make working with SPAs and ruby a delight. The work on the API is definitely getting versioned and updated accordingly. I do like how you added a namespace (at least) to help manage the complexity.

**Comment by BenMorganIO** [comment]
My current employer is launching their Spree site next week and then I'm off on vacation in an Indian desert (northern Rajastan) building either a school or a well. So, I'll be unavailable to implement this for a whole month. When I get back, the first things I'll be doing is (a) [Spree Book](https://leanpub.com/building-a-store-with-spree-ecommerce-for-ruby-on-rails), (b) site features, and (c) working on this major upgrade for 2 hours a day.

**Comment by hhff** [comment]
I used the 0.8.x track of AMS because it reflects the API of the 0.10 closest, and 0.10 wasn't ready when I started. 

@JDutil absolutely - there's definitely areas where spree_ams can improve - I actually still consider it "alpha".  I've pretty much just done the minimum work needed to get it working with an Ember storefront.  Areas that I'm aware of are:
- Auditing what data should be side loaded, and what data should be requested async
- Caching in general.  Caching in the AMS 0.8.x track is fairly primitive: Sideloaded models won't break the low level cache on update.
- The `api/ams/countries` call in particular sideloads the states for each country, it's the heaviest call of all and needs caching.
- Audit which endpoints actually need to exist.
- Auditing what data needs to be sent in the `meta` part of the payload.
- Standardize the way models are loaded "async".  Ember Data usually with make a request to a controller's index action, with a query parameter asking for specific IDs.
- I'd also love to move to AMS 0.10rc1.

I've got a docs harness in place via Apitome, but it needs a bunch of work before it's up to snuff: http://testing.spree-ember.com/api/ams/docs

So yeah!  That's just a start.  Would be good to organize a Google Hangout sometime soon and chat through next steps.  

I'd imagine a nice route would be to keep spree_ams separate from the spree codebase until we're happy that it's strong.  Actually merging it in will be fairly straight forward, but separating the complexity out as we iterate would be nice.

**Comment by hhff** [comment]
@BenMorganIO - totally.  The [order serializer](https://github.com/hhff/spree-ember/blob/master/packages/checkouts/addon/serializers/order.js) as part of the checkouts package encapsulates all of the specific logic for preparing an order for the Checkouts API.

Such an elegant way of managing that.  I've done it the wrong(tm) way so many times.

The other big win here was the [checkouts service](https://github.com/hhff/spree-ember/blob/master/packages/checkouts/app/services/checkouts.js).  It's a stateful abstraction on the current order's checkout state.  

It allows the Spree Ember developer to build a reactive single page checkout.  The checkouts state machine consolidates its state with the server at appropriate points, allowing the developer to program a UI to react to changes in the service.

It can also be reopened in the host application if the developer has modified Spree's checkout state machine.

:heart:

**Comment by BenMorganIO** [comment]
@hhff I don't feel like a spree_ams would be a good idea. I feel like things might get lost and it might be more effort to create a whole extension for that matter. But then again, this whole thing requires a lot of effort!

I think all the models should have endpoints if you ask me... That's my honest truthy opinion of this matter. A lot of work would therefore be required.

As per the checkout, I think that thing needs its own `ActiveModel::Model`. I was thinking: maybe a service? But, then I realized that its a stupid problem that can be easily solved with rails + MVC. The way forward, to me, is to create a single entry point for the api (`api/v2/checkout`) that would handle everything. If a developer wanted to handle it state specific, then POSTing the new state of an order might work better (`api/v2/orders/R123456789/states/address` from cart state, route would be structured `api/v2/orders/:order_number/states/:state`).

I've spent some time thinking about states with APIs and I don't think having a `next` endpoint is the worlds best idea... Letting the state become a param and then letting the resources state machine become a set of resources feels a bit better.

As per managing the states with checkout, well, that's just a model. That model could be used even in the `Spree::Api::V2::OrderStatesController` or whatever we end up calling it.

I think we need a lot of forward thinking to get this new API out the door and hope to never ever do a V3 API, just (hopefully) easily maintain V2.

**Comment by hhff** [comment]
Interesting!  I think a google hangout is in order!

Seeing as how deep you're into this - it might be worth you spinning up a Spree Ember app and combing through the requests, and checking it out how it works in detail.  It's actually pretty straight forward.

I've got a backend that's preconfigured with Spree 3.0 & Spree AMS over here, if you'd like an easy way to get up and going: https://github.com/hhff/spree-ember-testing

I opted for sideloading mainly to save requests.  That said - by having flat endpoints for every model, and making relationships asynchronous, caching is a lot easier, and many tiny requests is quite alot easier to handle from a complexity standpoint.

I agree re: the `next` endpoint.  Spree Ember doesn't use the `next` endpoint at all, but it does use `api/ams/checkouts` route as the entry point for modifying order state and detail - but that's more to be consistent with Spree's current Checkouts API.  

I think it's enough to send (attempted) `state` in the payload, rather than break it into routes & models.  Spree's FSM and Checkouts API is pretty solid in general, the hard part is translating that complexity to a clientside app and handling it in a sane way...

Regardless!  That's kinda in the weeds for now.  Im down to help however!

**Comment by yurifrl** [comment]
This is definitely the best thing that could happened to my day, I've been designing the api all wrong, this is such a great learning, thank you so much @hhff

**Comment by hhff** [comment]
So happy to be able to help @yurifrl !  Love the spree community :)

**Comment by MisinformedDNA** [comment]
I hate the way we use state machines. If I get to the confirmation state and then decide to add something to the cart, I don't really need to update my address and billing info again, though I may want to review my shipping options.

If the newly added item falls under the same shipping option, then it's not required to go to that state. The frontend could be changed to just keep using the selected option or it may want to bring them to that state anyway.

However, changing my shipping option does not mean that I need to go back to the payment state.

Amazon is a good example of this. You can quickly changes values here and there and they never make you go back through all the states. I think shipping options are always on the "confirm" page, so the user always see the latest.

**Comment by BenMorganIO** [comment]
I don't think you have to re update all the address information when an
order is in confirmation state. Please keep in mind that this RFC only
applies to the API and partially to core since its a dependency.

Reviewing shipping options and address specific information is tied
directly to your business. I don't think this is something spree should
concern or perfect. It's the job of your company to identify what work
works best for them and implement that.

I've never found the management of state in orders to require to update
previous information. I've found that the state machines provide some nice
leniency.

**Comment by hhff** [comment]
:+1: I think spree's state machine is tight, I can't imagine a cleaner way to write a customizable checkout flow!

**Comment by MisinformedDNA** [comment]
The state machine should be part of the frontend, not the backend.

One problem I've had with the current state machine is that if the user enters a CC and gets to the confirm state and then decides to go back to the payment state to change cards, I have to set the state of the checkout flow to be 'payment' state again, otherwise there are errors. The state machine can remain for those who use `spree_frontend`, but it should have no say in how the API is used.

My 2 cents.

**Comment by JDutil** [comment]
The state machine is part of core as it's at the model level not controller.  The state machine works fine, but the current implementation has become too intertwined and messy with callbacks.  Ultimately the state machine needs be to be ditched altogether for PORO service classes / interactors, however I don't foresee the community getting behind that project anytime soon as the consensus I see is everyone wanting to incrementally improve what we have for stability rather than fixing the root causes.  That is an entirely different discussion from the API though, and deserves it's own RFC.

@BenMorganIO I'm thinking we can merge this as active and just do subsequent PRs to improve/update further changes.

**Comment by hhff** [comment]
@JDutil @BenMorganIO - I've been holding off writing too much here as there's a ton of movement on the Ember Data / Ember 2.0 front.  I've wanted to wait until the dust settles on that first.

### **JSON API**

It's worth taking a look over [this new Ember blogpost about Ember Data 1.13.x](http://emberjs.com/blog/2015/06/18/ember-data-1-13-released.html).

The TL:DR of it is:  Follow the [JSON API](http://jsonapi.org/) spec closely, and Ember Data will _just work_.  It's the "happiest of happy paths" to Ember Data.

While we aren't necessarily trying to support Ember Data here, what I do like about this approach is that if we're building something for the greater community, it would be nice to be able to point them in the direction of the JSON API site, and say _"here, it works like this"_

I saw Steve Klabnik talk about JSON API a while back in NYC.  He made some great points about how JSON API will lower "bike shedding" discussions about API structure - but for me, the exciting part of it was what it would mean for Open Source APIs.

The nice things here are:
**Active Model Serializers will be bundled with Rails 5** and **Active Model Serializers has a JSON API Adapter!**

There's already work being done to support Backbone, Angular, GraphQL (React) and more.  Seems like an easy win - and we can even provide a config for the developer to opt out of it and use AMS's _flat_ adapter instead.

### **To Sideload or not to Sideload**

Ember 2.0 is going to make all Model Relationships Async by default.  In practice this means:

``` js
this.store.findRecord('orders', 'R382743');
```

Results in a promise / payload like this:

``` js
  // GET http://spree-api.com/api/v2/orders/R382743.json
  order: {
    line_item_ids: [1, 2]
  }
```

Then, in Ember code, retrieving those lineItems would then result in another promise / request cycle:

``` js
// Ember Code
order.get('lineItems').then(function(lineItems) {
  // GET http://spree-api.com/api/v2/line_items/1.json
  // GET http://spree-api.com/api/v2/line_items/2.json
  return lineItems;
});
```

**Following this strictly means we need an API endpoint for every model that we want to expose through the API, including product_properties, payment_profiles and more**

[spree_ams](https://github.com/hhff/spree_ams) does not do this.  Spree AMS manages to make do with 8 API endpoints to provide a fully functional Ember Frontend, by side-loading associated data.

**JSON API supports side loading data**, and Ember Data does also.  

In my opinion, async-everything is great for smaller apps, with ~10 data models or so.  I think it falls down when apps start getting bigger.

_Pros of async-everything_
- Caching is much easier (caching relationships is tricky)
- The UI can load associated data after the core data model is already present on the screen

_Cons of async-everything (Pros of sideloading)_
- Waaay less requests (in Spree's case)
- Like, 40+ endpoints (in Spree's case)
- JS doesn't like lots of parallel requests (laggy UI)

### **API Auth & CORS***

Something that doesn't jive too well with SPAs is the way API Auth works.  It's strange that every requester needs an API key just to view the products on the site via the API.  This is public information - and should not require a key.

This key is there for what I assume is a rounda-bout method of what CORS is designed for.  This is how we do it with [spree_ams](https://github.com/hhff/spree_ams/blob/3-0-alpha/lib/spree_ams/requestable.rb).  You can only request the `products` if you're permitted by the CORS policy.

Ideally, this RFC will make Spree more friendly to SPAs from an Authentication standpoint also.

I propose:
- We loosen up the API Key requirements on the API, in exchange for a strong CORS Policy.
- We move to an API token based authentication system for Users.  These should be loaded & authenticated against via request headers in the API base controller.  We can still use existing API Key mechanism here.

### **TLDR**

So my suggestions before we merge this are:
a) We add a note saying the Spree API will strictly follow the JSON API Spec

b) We decide either for or against sideloading, and if we opt into sideloading, we sketch out what Endpoints we'll need, and what sideloaded data they will provide.

c) We add a strong CORS policy, and officially support token based authentication through the API

Please let me know thoughts guys!!  :heart:

**Comment by JDutil** [comment]
My thoughts are:
1) I'm :+1: for JSON API Spec
2) I think async everything is a bad idea with the relationships in Spree
3) :+1: as well

I'm not doing much at all for Spree application development anymore though so it's more important to me that people actually using it day in day out are happy with the changes over my opinion on things.

**Comment by MisinformedDNA** [comment]
I am using the API v1 right now. I'm really glad that we are moving towards something that will be both easier to use and, from what I understand, more performant.

I've never used JSON API. It just reached stable version 1.0 on May 29, 2015, so that does make me a little worried.

I have used OData, which is in stable 4.0, been around since 2007 and became an OASIS standard in March 2014 and continues to be developed.

JSON API may be the next big thing and it appears that Rails is getting behind it, but it is still in it's nascent stages and that worries me some. But it might be just fine too.

More important is that the features we need are there. For some entities, I want less information. Like maybe I just want to set the shipping address and don't need the whole cart returned back to me. But sometimes I do want embedded entities, like when I do what the whole cart and all the various pieces.

OData allows you to say, I want this base object but also expand this associations. So you could get just the base cart, the cart with just the addresses, the cart with just shipping option, the cart with user data. It's super flexible for the consumer to grab as much or as little as possible. Maybe this is exists or is coming with JSON API? 

Since JSON API is being adopted by Rails, that seems like that would be a good choice, but I also think that building content negotiation into the API, so others could potentially build alternatives would be nice as well.

Also, there's a lot of talk about Ember in this thread. Ember should only be one data point. The Spree API needs to work nicely with all (at least, all popular) JS frameworks. It shouldn't be built for any one framework. There's still a lot of changes in the JS framework world and we don't want to make it easy for one framework only for the tides to change and then it's unnecessarily difficult to get the API working with other frameworks.

Overall, I think we are headed in a good direction and look forward to the new API!

**Comment by BenMorganIO** [comment]
@MisinformedDNA

Good thoughts on the JSON API.

The Ember discussion is more because its one of the few truly conventional JS frameworks. My feelings on Ember is that if we focus on making it work seamlessly, then Backbone, Knockout, and React integration will follow easily. Ember and React, in Vancouver, seem to be the big two frameworks right now too.

I think we're heading in a good direction too!

I'll get to updating this RFC with our new thoughts later today. Especially @hhff's.

**Comment by hhff** [comment]
**quick note:**
in JSON API you sideload relationships like this:

```
render @posts, include: ['authors', 'comments']
```

(Which is also how they are getting around the relationship caching problem - through fragment caching instead)

So I'd assume we'd sketch out what our endpoints include by default in our API controllers, and then if the developer decides they don't need certain models in their responses, they could decorate those controllers to only include what they want.

Amazing work @BenMorganIO - will hit you back on email tomorrow x

**Comment by BenMorganIO** [comment]
@hhff Can't wait. We should also chat about authentication as well. Especially token based authentication and letting Spree support non-session based auth. Token based auth is not apart of this RFC but something to work on after the V2 API work has started. This would make auth with Ember and React a lot easier.

**Comment by hhff** [comment]
Nice!  There's a super lightweight token auth working with Spree AMS - it only took a few lines of code to get Spree working with.  Will explain in greater detail on our chat!

**Comment by BenMorganIO** [comment]
@hhff awesome!

**Comment by MisinformedDNA** [comment]
In reference to sideloading relationships in JSON API, [this](http://jsonapi.org/format/#fetching-relationships) is what I was looking for!

``` http
GET /articles/1?include=comments.author HTTP/1.1
Accept: application/vnd.api+json
```

By doing this in the URL, as opposed to somewhere in the code, we keep things flexible.

Sometimes I just want the order state, well then I get just the order.
Maybe I'm checking out and want everything, well then I specify everything I want. 

Explicit instead of implicit. That means smaller and faster responses, by default.

**Comment by BenMorganIO** [comment]
@MisinformedDNA Great find! This would be a great feature to have and would work wonders with Spree!

**Comment by hhff** [comment]
:+1: LETS DO THIS!!!

**Comment by JDutil** [comment]
For those interested in following along @BenMorganIO open sourced his initial work, which people can try out or contribute to over at https://github.com/spree-contrib/spree_api_v2
